File: src/parser.rs
use crate::lexer::{LexerToken, LexerValue};
use crate::logger::Logger;
use crate::options::OptionsType;
use std::collections::HashMap;
use std::env;

pub struct Parser<'a> {
    envs: HashMap<String, String>,
    options: &'a OptionsType,
    tokens: Vec<LexerToken>,
    undefined_keys: Vec<String>,
    logger: Logger<'a>,
}

impl<'a> Parser<'a> {
    pub fn new(options: &'a OptionsType, tokens: Vec<LexerToken>) -> Self {
        let mut logger = Logger::new("Parser");
        logger.set_debug(&options.debug);

        Parser {
            envs: HashMap::new(),
            options,
            tokens,
            undefined_keys: options.required_envs.clone(),
            logger,
        }
    }

    pub fn get_envs(self) -> HashMap<String, String> {
        self.envs
    }

    pub fn parse_tokens(&mut self) {
        for token in &self.tokens {
            let mut value = String::new();
            let mut key = String::new();

            if !token.values.is_empty() {
                for value_token in &token.values {
                    if let Some(k) = &token.key {
                        key = String::from(k);
                    }
                    match value_token.token_type {
                        LexerValue::Interpolated => {
                            let mut interpolated_key = String::new();
                            if let Some(interpolated_value) = &value_token.value {
                                interpolated_key.push_str(interpolated_value);
                            }

                            if let Ok(env_val) = env::var(&interpolated_key) {
                                value.push_str(env_val.as_str());
                            } else if let Some(env_map_val) = self.envs.get(&interpolated_key) {
                                value.push_str(env_map_val);
                            } else if self.options.debug {
                                self.logger.warn(
                                    format!(
                                        "found a key {:?} that contains an invalid interpolated variable: {:?} ({}:{}:{}). Unable to locate a value that corresponds to this key. Skipping.",
                                        key, interpolated_key, token.file, value_token.line, value_token.byte
                                    )
                                );
                            }
                        }
                        LexerValue::Comment => {
                            if let Some(val) = &value_token.value {
                                self.logger.debug(format!(
                                    "parsed the following comment {:?} ({}:{}:{}). Skipping.",
                                    val, token.file, value_token.line, value_token.byte
                                ));
                            }
                        }
                        _ => {
                            if let Some(val) = &value_token.value {
                                value.push_str(val.as_str());
                            }
                        }
                    }
                }

                if !key.is_empty() {
                    self.envs.insert(key.clone(), value);
                    if !self.undefined_keys.is_empty() {
                        self.undefined_keys.retain(|k| *k != key);
                    }
                }
            }
        }

        if self.envs.is_empty() {
            self.logger.fatal(String::from(
                "Unable to parse any ENVs! Please ensure the provided .env files are not empty.)",
            ));
        }

        if !self.undefined_keys.is_empty() {
            self.logger.fatal(
                format!(
                    "found that the following ENVs were marked as required: {:?}, but they are undefined after all of the ENVs were parsed. Perhaps you forgot to include them in an .env file or an API project's environment?", 
                    self.undefined_keys
                )
             );
        }

        self.logger.debug(format!(
            "generated the following env map...\n{:#?}",
            self.envs
        ));
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::lexer::Lexer;
    use crate::options::Options;
    use std::process::{Command, Stdio};
    use std::sync::OnceLock;

    fn get_envs() -> &'static HashMap<String, String> {
        static ENVS: OnceLock<HashMap<String, String>> = OnceLock::new();
        ENVS.get_or_init(|| {
            let argv = vec![
                String::from("nvi"),
                String::from("--debug"),
                String::from("--directory"),
                String::from("envs"),
                String::from("--files"),
                String::from(".env"),
                String::from("base.env"),
                String::from("reference.env"),
            ];
            let options = Options::new(argv);

            let mut lexer = Lexer::new(&options);
            lexer.tokenize();

            let mut parser = Parser::new(&options, lexer.get_tokens());
            parser.parse_tokens();

            parser.get_envs()
        })
    }

    fn get_env(key: &str) -> String {
        let envs = get_envs();

        match envs.get(key) {
            Some(v) => String::from(v),
            None => String::new(),
        }
    }

    #[test]
    fn basic_env() {
        assert_eq!(get_env("BASIC_ENV"), String::from("true"));
    }

    #[test]
    fn base_env() {
        assert_eq!(get_env("BASE"), String::from("hello"));
    }

    #[test]
    fn ref_base_env() {
        assert_eq!(get_env("REFERENCE"), String::from("hello"));
    }

    #[test]
    fn multiline_env() {
        assert_eq!(
            get_env("MULTI_LINE_KEY"),
            String::from("ssh-rsa BBBBPl1P1AD+jk/3Lf3Dw== test@example.com")
        );
    }

    #[test]
    fn no_env() {
        assert_eq!(get_env("NO_VALUE"), String::from(""));
    }

    #[test]
    fn hash_env() {
        assert_eq!(get_env("HASH_KEYS"), String::from("1#2#3#4#"));
    }

    #[test]
    fn dollars_env() {
        assert_eq!(get_env("JUST_DOLLARS"), String::from("$$$$$"));
    }

    #[test]
    fn braces_env() {
        assert_eq!(get_env("JUST_BRACES"), String::from("{{{{}}}}"));
    }

    #[test]
    fn spaces_env() {
        assert_eq!(get_env("JUST_SPACES"), String::from("          "));
    }

    #[test]
    fn sentence_env() {
        assert_eq!(
            get_env("SENTENCE"),
            String::from(
                "chat gippity is just a junior engineer that copies/pastes from StackOverflow"
            )
        );
    }

    #[test]
    fn interpolate_from_system_env() {
        assert_eq!(get_env("INTERP_ENV_FROM_SYS").len() > 0, true);
    }

    #[test]
    fn empty_interpolated_env() {
        assert_eq!(get_env("EMPTY_INTERP_KEY"), String::from("abc123"));
    }

    #[test]
    fn single_quotes_env() {
        assert_eq!(
            get_env("SINGLE_QUOTES_SPACES"),
            String::from("'  SINGLE QUOTES  '")
        );
    }

    #[test]
    fn double_quotes_env() {
        assert_eq!(
            get_env("DOUBLE_QUOTES_SPACES"),
            String::from("\"  DOUBLE QUOTES  \"")
        );
    }

    #[test]
    fn quotes_env() {
        assert_eq!(get_env("QUOTES"), String::from("sad\"wow\"bak"));
    }

    #[test]
    fn invalid_single_quotes_env() {
        assert_eq!(get_env("INVALID_SINGLE_QUOTES"), String::from("bad\'dq"));
    }

    #[test]
    fn invalid_double_quotes_env() {
        assert_eq!(get_env("INVALID_DOUBLE_QUOTES"), String::from("bad\"dq"));
    }

    #[test]
    fn invalid_template_literal_env() {
        assert_eq!(get_env("INVALID_TEMPLATE_LITERAL"), String::from("bad`as"));
    }

    #[test]
    fn invalid_mix_quotes_env() {
        assert_eq!(get_env("INVALID_MIX_QUOTES"), String::from("bad\"'`mq"));
    }

    #[test]
    fn empty_single_quotes_env() {
        assert_eq!(get_env("EMPTY_SINGLE_QUOTES"), String::from("''"));
    }

    #[test]
    fn empty_double_quotes_env() {
        assert_eq!(get_env("EMPTY_DOUBLE_QUOTES"), String::from("\"\""));
    }

    #[test]
    fn empty_template_literal_env() {
        assert_eq!(get_env("EMPTY_TEMPLATE_LITERAL"), String::from("``"));
    }

    #[test]
    fn multiple_interpolated_env() {
        assert_eq!(
            get_env("MONGOLAB_URI"),
            String::from("mongodb://root:password@abcd1234.mongolab.com:12345/localhost")
        );
    }

    #[test]
    fn prints_empty_env_error() {
        match Command::new("./target/debug/nvi")
            .args(["--directory", "envs", "--files", "comment.env"])
            .stdin(Stdio::null())
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .spawn()
        {
            Ok(c) => {
                let output = c.wait_with_output().expect("Failed to read command output");

                assert!(!output.status.success());

                let stderr = String::from_utf8_lossy(&output.stderr);

                assert!(
                    stderr.contains("Unable to parse any ENVs! Please ensure the provided .env files are not empty."),
                );
            }
            Err(err) => {
                panic!("Failed to run prints_empty_env_error test. Reason: {err}");
            }
        }
    }

    #[test]
    fn prints_required_env_error() {
        match Command::new("./target/debug/nvi")
            .args([
                "--directory",
                "envs",
                "--files",
                "base.env",
                "--required",
                "KEY1",
            ])
            .stdin(Stdio::null())
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .spawn()
        {
            Ok(c) => {
                let output = c.wait_with_output().expect("Failed to read command output");

                assert!(!output.status.success());

                let stderr = String::from_utf8_lossy(&output.stderr);

                assert!(
                    stderr.contains("found that the following ENVs were marked as required: [\"KEY1\"], but they are undefined"),
                );
            }
            Err(err) => {
                panic!("Failed to run prints_required_env_error test. Reason: {err}");
            }
        }
    }
}

File: src/parser.zig
const std = @import("std");
const tty = @import("tty.zig");
const arg = @import("arg.zig");
const tk = @import("tokenizer.zig");
const fmt = @import("formatter.zig");

const Io = std.Io;
const mem = std.mem;
const Environ = std.process.Environ;

pub const Parser = struct {
    environ: *Environ.Map,
    alloc: mem.Allocator,
    args: *const arg.Arg,
    tokens: *const std.ArrayList(tk.Token),
    logger: *Io.Writer,
    envs: std.StringArrayHashMapUnmanaged([]const u8) = .empty,

    pub fn run(self: *Parser) !void {
        for (self.tokens.items) |tkn| {
            var value: std.ArrayList(u8) = .empty;
            defer value.deinit(self.alloc);

            // a commented token contributes nothing and is skipped whole
            if (try self.resolveValue(tkn, &value)) continue;

            if (value.items.len == 0) {
                try self.warnUndefined(tkn);
                continue;
            }

            try self.envs.put(self.alloc, tkn.key.?, try value.toOwnedSlice(self.alloc));
        }

        if (self.envs.count() == 0) {
            try self.logger.writeAll(
                tty.red ++ "error:" ++ tty.reset ++ " Unable to parse any ENVs! Please ensure the provided .env files are not empty.\n",
            );
            return error.NoParsedEnvs;
        }

        if (self.args.debug) try self.printListing();

        try self.checkRequired();
    }

    fn resolveValue(self: *Parser, tkn: tk.Token, value: *std.ArrayList(u8)) !bool {
        for (tkn.values.items) |value_token| {
            switch (value_token.kind) {
                .interpolated => {
                    const interpolated_key = value_token.value;

                    if (self.environ.get(interpolated_key)) |v| {
                        try value.appendSlice(self.alloc, v);
                    } else if (self.envs.get(interpolated_key)) |v| {
                        try value.appendSlice(self.alloc, v);
                    } else if (self.args.debug) {
                        try self.logger.print(
                            tty.yellow ++ "warning:" ++ tty.reset ++ " The " ++ tty.bold_yellow ++ "{s}" ++ tty.reset,
                            .{tkn.key.?},
                        );
                        try self.logger.print(
                            " key contains an interpolated key variable " ++ tty.bold_yellow ++ "{s}" ++ tty.reset,
                            .{interpolated_key},
                        );
                        try self.logger.print(
                            " ({s}:{d}:{d}) that is not defined; skipping.\n",
                            .{ tkn.file, value_token.line, value_token.byte },
                        );
                    }
                },
                .commented => {
                    if (self.args.debug) {
                        try self.logger.print(
                            tty.cyan ++ "info:" ++ tty.reset ++ " Parsed a comment:" ++ tty.cyan ++ " {s} " ++ tty.reset,
                            .{value_token.value},
                        );
                        try self.logger.print(
                            "({s}:{d}:{d}); skipping.\n",
                            .{ tkn.file, value_token.line, value_token.byte },
                        );
                    }
                    return true;
                },
                else => try value.appendSlice(self.alloc, value_token.value),
            }
        }

        return false;
    }

    fn warnUndefined(self: *Parser, tkn: tk.Token) !void {
        const first = tkn.values.items[0];
        const key_str = if (tkn.key != null) tkn.key.? else "(undefined)";

        try self.logger.print(
            tty.yellow ++ "warning:" ++ tty.reset ++ " The " ++ tty.bold_yellow ++ "{s}" ++ tty.reset,
            .{key_str},
        );
        try self.logger.writeAll(" key is an " ++ tty.bold_yellow ++ "undefined" ++ tty.reset);
        try self.logger.print(
            " value ({s}:{d}:{d}); skipping.\n",
            .{ tkn.file, first.line, first.byte },
        );
    }

    fn printListing(self: *Parser) !void {
        try self.logger.print(
            "\n" ++ tty.cyan ++ "info:" ++ tty.reset ++ " The following {d} env(s) were parsed and will be emitted...\n",
            .{self.envs.count()},
        );

        var it = self.envs.iterator();
        while (it.next()) |entry| {
            try self.logger.writeAll("  • ");
            try fmt.printPair(self.logger, self.args.format, entry.key_ptr.*, entry.value_ptr.*);
            try self.logger.writeByte('\n');
        }

        try self.logger.writeByte('\n');
    }

    fn checkRequired(self: *Parser) !void {
        if (self.args.required.items.len == 0) return;

        var missing: std.ArrayList([]const u8) = .empty;
        defer missing.deinit(self.alloc);

        for (self.args.required.items) |req| {
            if (!self.envs.contains(req)) try missing.append(self.alloc, req);
        }

        if (missing.items.len == 0) return;

        try self.logger.writeAll(tty.red ++ "error" ++ tty.reset ++ ": The following ENV keys were marked as required, but weren't set after parsing: ");

        for (missing.items) |key| {
            try self.logger.print("\n   • " ++ tty.bold_red ++ "{s}" ++ tty.reset, .{key});
        }

        try self.logger.writeAll("\nThese keys are either set at runtime by a process or were supposed to be set in an .env file.");
        try self.logger.writeAll(" You can either ignore them (see help for more info) or include them in an .env file.\n");

        return error.MissingRequiredEnvs;
    }
};
